armory/art_experimental/attacks/sweep.py

```diff
@@ -0,0 +1,228 @@
+            self.num_search_points = len(list(self.sweep_kwargs.values())[0])
+        else:
+            self.num_search_points = len(list(self.sweep_generate_kwargs.values())[0])
+
+        for i in range(self.num_search_points):
+            subattack_i_kwargs = kwargs.copy()
+            for sweep_attack_kwarg in self.sweep_kwargs:
+                subattack_i_kwargs[sweep_attack_kwarg] = self.sweep_kwargs[
+                    sweep_attack_kwarg
+                ][i]
+            subattack_i = attack_fn(estimator, **subattack_i_kwargs)
+            self.attacks.append(subattack_i)
+
+    def generate(self, x, y=None, **kwargs):
+        if x.shape[0] != 1:
+            raise ValueError("Dataset batch_size should be set to 1")
+        if y is None:
+            raise ValueError(
+                "This attack requires given labels. Ensure that attack['use_label'] is set to True."
+            )
+
+        if len(kwargs.keys() & self.sweep_generate_kwargs.keys()) > 0:
+            logger.warning(
+                f"The following kwargs were set in both attack_config['generate_kwargs'] and "
+                f"attack_config['sweep_params']['generate_kwargs']: "
+                f"{list(kwargs.keys() & self.sweep_generate_kwargs.keys())}. Values specified "
+                f"in attack_config['generate_kwargs'] will be ignored."
+            )
+
+        y_pred = self._estimator.predict(x)
+        if not self._is_robust(y, y_pred):
+            logger.info(
+                f"Estimator is not robust to original x as measured with metric "
+                f"function {self.metric_fn.__name__} and threshold "
+                f"{self.metric_threshold}. Returning original x."
+            )
+            return x
+
+        i_min = 0
+        i_max = self.num_search_points
+        x_best = None
+        while i_min < i_max:
+            i_mid = (i_min + i_max) // 2
+            attack_kwargs = {k: v[i_mid] for k, v in self.sweep_kwargs.items()}
+            generate_kwargs = {
+                k: v[i_mid] for k, v in self.sweep_generate_kwargs.items()
+            }
+            kwargs.update(generate_kwargs)
+            x_adv = self.attacks[i_mid].generate(x, y, **kwargs)
+            y_pred = self._estimator.predict(x_adv)
+            if not self._is_robust(y, y_pred):
+                # attack success
+                logger.info(
+                    f"Success with kwargs {attack_kwargs} and generate_kwargs {generate_kwargs}"
+                )
+                x_best = x_adv
+                best_attack_kwargs = attack_kwargs
+                best_generate_kwargs = generate_kwargs
+                i_max = i_mid
+            else:
+                # attack failure
+                i_min = i_mid + 1
+                logger.info(
+                    f"Failure with kwargs {attack_kwargs} and generate_kwargs {generate_kwargs}"
+                )
+
+        if x_best is None:
+            logger.info("Attack failed. Returning original x.")
```

Comment:
It may be useful to be able to extract the actual attack_kwargs and generate_kwargs at a per sample level -- it would be slightly easer if there were a common prefix for successful and failing attacks, so that a string could be grep'ed for. This might be useful to keep in the results json as well, but that would likely break the abstraction of the attack...

Reply:
now using common prefix `"Sweep attack concluded."`